module/Application/test/ModuleTest.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ApplicationTest;
+
+use Application\Module;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \Application\Module;
+ */
+class ModuleTest extends TestCase
```

Comment:
This class can be final and if we have a test for the module class, then we also need a test for the index controller in this module.

Reply:
Controllers are not testable in isolation because their behavior depends on what is setup for the by the whole mvc application. I never bother testing controllers directly.

Reply:
I see no reason to make test class final. What is your reasoning?

Reply:
> Controllers are not testable in isolation because their behavior depends on what is setup for the by the whole mvc application. I never bother testing controllers directly.

But this is exactly what is explained in the laminas-test documentation. So the question is, why do we need laminas-here here at all?

> I see no reason to make test class final. What is your reasoning?

Here I only follow the tests in other repositories. See laminas-form or laminas-servicemanager as example.

Reply:
> But this is exactly what is explained in the laminas-test documentation. So the question is, why do we need laminas-here here at all?

I think I misunderstood you when you asked about index controller test. 
laminas-test is used for existing index controller integration test which is included as an example. 
 
It is still not too reliable because behavior changes at runtime with events and listeners but it is better than no tests or isolated unit test.
Only e2e tests are approaching required assurance but they also incredibly fragile.

> Here I only follow the tests in other repositories. See laminas-form or laminas-servicemanager as example.

I do not recall this established as common practice for the framework so it is up to personal taste of maintainer at this point. Since tests are not available as a dependency and there is no implied guarantee of compatibility or stability the final modifier is moot.
I do not care either way.

Reply:
@froschdesign do you still want me to add final to all tests classes?

Reply:
I just wanted to point out that almost all tests have been marked as final for some time now.
Didn't want to start a discussion.